File: 백준/Silver/10451. 순열 사이클/순열 사이클.py
import sys
input = sys.stdin.read

def dfs(node, visited, permutation):
    while not visited[node]:
        visited[node] = True
        node = permutation[node]

def count_cycles(n, permutation):
    visited = [False] * (n + 1)
    cycles = 0

    for i in range(1, n + 1):
        if not visited[i]:
            dfs(i, visited, permutation)
            cycles += 1
    
    return cycles

data = input().split()
t = int(data[0])
idx = 1
results = []

for _ in range(t):
    n = int(data[idx])
    permutation = [0] + list(map(int, data[idx + 1: idx + 1 + n]))
    idx += n + 1
    results.append(count_cycles(n, permutation))

print("\n".join(map(str, results)))

File: 백준/Silver/10451. 순열 사이클/순열 사이클.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;

public class Main {

    static void dfs(int node, boolean[] visited, int[] permutation) {
        while (!visited[node]) {
            visited[node] = true;
            node = permutation[node];
        }
    }

    static int countCycles(int n, int[] permutation) {
        boolean[] visited = new boolean[n + 1];
        int cycles = 0;
        for (int i = 1; i <= n; i++) {
            if (!visited[i]) {
                dfs(i, visited, permutation);
                cycles++;
            }
        }
        return cycles;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] allInput = br.lines().toArray(String[]::new);
        int t = Integer.parseInt(allInput[0]);
        StringBuilder sb = new StringBuilder();

        int idx = 1;
        for (int test = 0; test < t; test++) {
            int n = Integer.parseInt(allInput[idx]);
            int[] permutation = new int[n + 1];
            String[] nums = allInput[idx + 1].split(" ");
            for (int i = 1; i <= n; i++) {
                permutation[i] = Integer.parseInt(nums[i - 1]);
            }
            idx += 2;
            sb.append(countCycles(n, permutation)).append("\n");
        }

        System.out.print(sb);
    }
}